### PR TITLE
objmode complex type spelling

### PR DESCRIPTION
**Pull request by sklam**
Fixes https://github.com/numba/numba/issues/4401
Fixes https://github.com/numba/numba/issues/3372
Fixes https://github.com/numba/numba/issues/3356

- [x] ~allow the use of global name for typename~ (removed; prefer using non-str reference to global and cell vars)
- [x] allow the use of cell vars
- [x] allow the use of global variable

### CODE REVIEW

File: numba/tests/test_withlifting.py [hunk 1]
```diff
@@ -800,6 +803,34 @@ def f():
 
         self.assertEqual(f(), 1 + 3)
 
+    def test_objmode_gv_name(self):
+        @njit
+        def global_name():
+            with objmode(val='gv_type'):
```
Comment (eric-wieser) [inline numba/tests/test_withlifting.py:809]: Shouldn't this be the following?
```suggestion
            with objmode(val=gv_type):
```

**Comment by gmarkall** [comment]
Is this now ready for review?

File: numba/tests/test_withlifting.py (hunk 1 above)
Comment (sklam) [reply numba/tests/test_withlifting.py:804]: Testing the use of global variable is done in `test_objmode_gv_variable`.
But now i wonder if we actually need the support for spelling type as string at all.

Comment (sklam) [reply numba/tests/test_withlifting.py]: Removed the support of using `str` to reference global variables.

File: numba/tests/test_withlifting.py [hunk 2]
```diff
@@ -795,6 +798,85 @@ def f():
...
+    def test_objmode_closure_type_in_overload(self):
+        def foo():
+            pass
+
+        @overload(foo)
+        def foo_overload():
+            shrubbery = types.float64[:]
+            def impl():
+                with objmode(out=shrubbery):
+                    out = np.ones(10).astype(np.float64)
```
Comment (stuartarchibald) [inline numba/tests/test_withlifting.py:829]: aren't these already float64?

File: numba/tests/test_withlifting.py [hunk 3]
```diff
@@ -795,6 +798,85 @@ def f():
...
+    def test_objmode_closure_type_in_overload_error(self):
+        def foo():
+            pass
+
+        @overload(foo)
+        def foo_overload():
+            shrubbery = types.float64[:]
+            def impl():
+                with objmode(out=shrubbery):
+                    out = np.ones(10).astype(np.float64)
```
Comment (stuartarchibald) [inline numba/tests/test_withlifting.py:846]: same here, RE float64

File: numba/tests/test_withlifting.py [hunk 4]
```diff
@@ -795,6 +798,85 @@ def f():
 
         self.assertEqual(f(), 1 + 3)
 
+    def test_objmode_gv_variable(self):
```
Comment (stuartarchibald) [inline numba/tests/test_withlifting.py:801]: might be good to check that the dtypes of the return values are as expected throughout these new tests?

File: numba/tests/test_withlifting.py [hunk 5]
```diff
@@ -795,6 +798,85 @@ def f():
...
+                t1 = 793856.5
+                t2 = t1         # to observe truncation
+                t3 = np.arange(5).astype(np.int32)
+            return t1, t2, t3
+
+        t1, t2, t3 = foo()
+        self.assertPreciseEqual(t1, 793856.5)
+        self.assertPreciseEqual(t2, 793856)
+        self.assertPreciseEqual(t3, np.arange(5).astype(np.int32))
+
```
Comment (stuartarchibald) [inline numba/tests/test_withlifting.py:879]: I've also tested somewhat awkward things like:
```python
from numba import njit, types, objmode
from numba.experimental import jitclass
import numpy as np
from numba.extending import overload


import numpy as np
from numba import int32, float32    # import the types
from numba.experimental import jitclass

spec = [
    ('value', int32),               # a simple scalar field
    ('array', float32[:]),          # an array field
]

@jitclass(spec)
class Bag(object):
    def __init__(self, value):
        self.value = value
        self.array = np.zeros(value, dtype=np.float32)

    @property
    def size(self):
        return self.array.size

    def increment(self, val):
        for i in range(self.size):
            self.array[i] += val
        return self.array

    @staticmethod
    def add(x, y):
        return x + y

n = 21
mybag = Bag(n)


def foo():
    pass

@overload(foo)
def foo_overload():
    shrubbery = mybag._numba_type_
    def impl():
        with objmode(out=shrubbery):
            out = Bag(123)
        return out
    return impl

@njit
def bar():
    return foo()

z = bar()
print(z, type(z))
print(z.add(2, 3))
```
which work fine.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, this is a good usability improvement. There's a few minor things to resolve but then it's good to merge. Did you want to update any docs to go alongside this change?

File: numba/tests/test_withlifting.py (hunk 4 above)
Comment (sklam) [reply numba/tests/test_withlifting.py:801]: i must have accidentally deleted somes lines. The function is never exercised.

File: numba/tests/test_withlifting.py (hunk 5 above)
Comment (sklam) [reply numba/tests/test_withlifting.py:887]: I added a test with your code

File: numba/tests/test_withlifting.py (hunk 2 above)
Comment (sklam) [reply numba/tests/test_withlifting.py]: switched to `ones` to `arange`

File: numba/tests/test_withlifting.py (hunk 3 above)
Comment (sklam) [reply numba/tests/test_withlifting.py]: same fix as above

File: numba/core/withcontexts.py [hunk 6]
```diff
@@ -127,29 +127,38 @@ class _ObjModeContextType(WithContext):
 
     Use this as a function that takes keyword arguments only.
     The argument names must correspond to the output variables from the
-    with-block.  Their respective values are strings representing the expected
-    types.  When exiting the with-context, the output variables are cast
+    with-block.  Their respective values can be:
+
+    1. strings representing the expected types; i.e. ``"float32"``.
+    2. global or closure variables referring to the expected type.
```
Comment (stuartarchibald) [inline numba/core/withcontexts.py:133]:
```suggestion
    2. global or free variables referring to the expected type.
```

File: numba/core/withcontexts.py [hunk 7]
```diff
@@ -127,29 +127,38 @@ class _ObjModeContextType(WithContext):
...
+       The variables are read at compile time.
+
+    When exiting the with-context, the output variables are casted
```
Comment (stuartarchibald) [inline numba/core/withcontexts.py:136]:
```suggestion
    When exiting the with-context, the output variables are converted
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the fixes, couple of wording suggestions else looks good. Thanks for adding the additional tests too.

File: numba/core/withcontexts.py (hunk 6 above)
Comment (sklam) [reply numba/core/withcontexts.py:133]: I especially want to mention that the "closure variables" must be bounded at compile time.

Comment (sklam) [reply numba/core/withcontexts.py:133]: So it might have to be "compile-time bounded global or nonlocal variables ..."

Comment (stuartarchibald) [reply numba/core/withcontexts.py:133]: Ah yes, from the code, `free` is too much, the latter sentence suggestion looks good.

File: numba/core/withcontexts.py [hunk 8]
```diff
@@ -127,29 +127,38 @@ class _ObjModeContextType(WithContext):
 
     Use this as a function that takes keyword arguments only.
     The argument names must correspond to the output variables from the
-    with-block.  Their respective values are strings representing the expected
-    types.  When exiting the with-context, the output variables are cast
+    with-block.  Their respective values can be:
+
+    1. strings representing the expected types; i.e. ``"float32"``.
+    2. compile-time bounded global or nonlocal variables referring to the
```
Comment (stuartarchibald) [inline numba/core/withcontexts.py:133]:
```suggestion
    2. compile-time bound global or nonlocal variables referring to the
```

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes!

**Pull request merged**